### PR TITLE
Add repo filtering to Gerrit code host connections

### PR DESCRIPTION
**Pull request by pjlast**
Adds a Gerrit code host config option to specify specific projects on Gerrit.

## Test plan

Add unit tests

### CODE REVIEW

File: internal/repos/gerrit.go [hunk 1]
```diff
@@ -22,10 +22,12 @@ import (
 // in Sourcegraph via the external services configuration.
 type GerritSource struct {
 	svc       *types.ExternalService
+	config    *schema.GerritConnection
 	cli       *gerrit.Client
 	serviceID string
 	perPage   int
 	private   bool
+	syncRepos map[string]struct{}
```
Comment (sashaostrikov) [inline internal/repos/gerrit.go:30]: This is unused, is it intentional?

File: internal/repos/gerrit.go [hunk 2]
```diff
@@ -79,6 +82,10 @@ func (s *GerritSource) CheckConnection(ctx context.Context) error {
 
 // ListRepos returns all Gerrit repositories configured with this GerritSource's config.
 func (s *GerritSource) ListRepos(ctx context.Context, results chan SourceResult) {
+	allowedProjects := make(map[string]struct{})
+	for _, project := range s.config.Repos {
```
Comment (sashaostrikov) [inline internal/repos/gerrit.go:86]: we put `repos` in `allowedProjects` 😁 
should we make this naming consistent?

**Review by sashaostrikov — APPROVED** [review]
LGTM!
Added a couple of questions about naming and unused property.

**Comment by sg-e2e-regression-test-bob** [comment]
## Bundle size report 📦

| Initial size | Total size | Async size | Modules |
| --- | --- | --- | --- |
| 0.00% (0.00 kb) | 0.02% (+2.88 kb) | 0.02% (+2.88 kb) | 0.00% (0) |

Look at the [Statoscope report](https://console.cloud.google.com/storage/browser/_details/sourcegraph_reports/statoscope-reports/pjlast/gerrit-repo-filtering/compare-report.html;tab=live_object) for a full comparison between the commits 9fe5c18 and 5af62f4 or [learn more](https://docs.sourcegraph.com/dev/how-to/testing#bundlesize).


<details>
<summary>Open explanation</summary>

- `Initial size` is the size of the initial bundle (the one that is loaded when you open the page)
- `Total size` is the size of the initial bundle + all the async loaded chunks
- `Async size` is the size of all the async loaded chunks
- `Modules` is the number of modules in the initial bundle
</details>

File: internal/repos/gerrit.go (hunk 1 above)
Comment (pjlast) [reply internal/repos/gerrit.go]: Sorry, marked the PR as ready before it actually was ready :D 

Functionality turned out a bit different than I anticipated, this is unused now. Removed ✅

File: internal/repos/gerrit.go (hunk 2 above)
Comment (pjlast) [reply internal/repos/gerrit.go]: Good point! Renaming to Projects because that's what Gerrit calls them

File: internal/repos/gerrit_test.go [hunk 3]
```diff
@@ -2,39 +2,83 @@ package repos
...
+
+		ctx := context.Background()
+		src, err := NewGerritSource(ctx, svc, cf)
+		if err != nil {
+			t.Fatal(err)
+		}
+
+		src.perPage = 25
+
+		repos, err := listAll(context.Background(), src)
```
Comment (sashaostrikov) [inline internal/repos/gerrit_test.go:37]: 🚨 Variable usage police here 🚨
```suggestion
		repos, err := listAll(ctx, src)
```

File: internal/repos/gerrit_test.go [hunk 4]
```diff
@@ -2,39 +2,83 @@ package repos
...
```
Comment (sashaostrikov) [inline internal/repos/gerrit_test.go:69]: 🚨 Variable usage police here 🚨
```suggestion
		repos, err := listAll(ctx, src)
```

File: internal/repos/gerrit_test.go [hunk 5]
```diff
@@ -2,39 +2,83 @@ package repos
...
+		}
+
+		testutil.AssertGolden(t, "testdata/sources/GERRIT/"+t.Name(), update(t.Name()), repos)
+	})
+
+	t.Run("with filtering", func(t *testing.T) {
+		conf := &schema.GerritConnection{
+			Url: "https://gerrit-review.googlesource.com",
+			Projects: []string{
+				"apps/reviewit",
```
Comment (sashaostrikov) [inline internal/repos/gerrit_test.go:49]: Reviewed, thanks!

File: internal/repos/gerrit_test.go [hunk 6]
```diff
@@ -2,39 +2,83 @@ package repos
...
+		defer save(t)
+
+		svc := &types.ExternalService{
+			Kind:   extsvc.KindGerrit,
+			Config: extsvc.NewUnencryptedConfig(marshalJSON(t, conf)),
+		}
+
+		ctx := context.Background()
+		src, err := NewGerritSource(ctx, svc, cf)
+		if err != nil {
```
Comment (sashaostrikov) [inline internal/repos/gerrit_test.go:31]: nit: such checks can be inlined with:
```suggestion
		require.NoError(t, err)
```

**Review by sashaostrikov — APPROVED** [review]
LGTM [2]! Noticed all naming fixes, left a couple of comments in tests.

File: schema/gerrit.schema.json [hunk 7]
```diff
@@ -29,6 +29,15 @@
       "type": "string",
       "minLength": 1
     },
+    "projects": {
```
Comment (kopancek) [inline schema/gerrit.schema.json:32]: Should we add docs about this property to Gerrit docs on setting it up?

**Review by kopancek — APPROVED** [review]

File: schema/gerrit.schema.json (hunk 7 above)
Comment (kopancek) [reply schema/gerrit.schema.json:32]: Nevermind, see the other PR now 🤦

Comment (pjlast) [reply schema/gerrit.schema.json:32]: It's mentioned here: https://github.com/sourcegraph/sourcegraph/pull/47019/files#diff-bc533a32a2f750af71c0ad4a2e623f66a2ac5f22b3e0c189f8d3dcc171debb42R18

**Pull request merged**